### PR TITLE
sriov, Fix race condition between LoginToFedora and cloud init

### PR DESCRIPTION
**Pull request by oshoval**
There is a potential race between e2e tests `LoginToFedora`,
and Cloud-init which reconfigure the password.

The flow is as follows:
1. e2e spins VM and wait for it to be 'Running'
2. e2e tries to Login to the VM via the console.
3. e2e waits for Agent connected event.

As part of the VM spinning this flow occurs:
A. The VM has Cloud-init UserData, which cause the Cloud-init to reconfig the VM
according the default distro configuration which reset the fedora user to not have a password.
And then runs the UserData commands which adds a password.

In case A started but didn't finish before 2 (login attempt), there will be a race that will cause the login
to fail.
This race can cause the login to eventually fail, in case the retires are exhausted, or in case the console
expecter is going out of sync with the acutal console output (the login prompt is not idempotent when it happens).

In order to fix the race and keep the code as simple as possible,
this PR does the following:

* Use a VM without Cloud-init module which reconfigure the fedora user
to its default values.
* Remove the setting of the password via UserData, as it doesn't needed anymore,
since the password now isn't reset by the Cloud-init default config.

Image built by
kubevirt/kubevirtci#556

```release-note
None
```

### CODE REVIEW

File: WORKSPACE
Comment (oshoval) [inline WORKSPACE:328]: Once the PR is approved, will push to an official repo

File: tests/libvmi/factory.go
Comment (davidvossel) [inline tests/libvmi/factory.go:55]: why does userdata need to be defined here if the script doesn't do anything?

**Review by davidvossel — COMMENTED** [review]
so, with this new sriov image, is the password preconfigured?

If so, how are we handling tests that need to ensure that the cloud-init script has executed before peforming the console login... for example, if a cloud-init script is setting something up for a test, and the fedora login can happen before the cloud-init script executes, then the test might kick off before the initial setup has completed.

I think the fedora password change in the cloud init was actually acting as sort of a gating mechanism to ensure we couldn't log in until the cloud-init is executed. We might not have planned for that to be the case, but that could be a dependency that has developed over time.

File: tests/libvmi/factory.go
Comment (oshoval) [reply tests/libvmi/factory.go:55]: It doesn't need to be, it just so it will be easy to add it, i can remove it (verified it works in all options)

**Comment by oshoval** [comment]
> so, with this new sriov image, is the password preconfigured?
> 

Yes

> If so, how are we handling tests that need to ensure that the cloud-init script has executed before peforming the console login... for example, if a cloud-init script is setting something up for a test, and the fedora login can happen before the cloud-init script executes, then the test might kick off before the initial setup has completed.
> 

I am planning on doing this specific commit that will wait for cloud-init to finish https://github.com/kubevirt/kubevirt/pull/5072/commits/8677da2582d0ebc1d75d9846e6e1fff8529bf811
just thought its not part of this PR
(need to add there that it will also check cloud-init is finished and finished successfully)

> I think the fedora password change in the cloud init was actually acting as sort of a gating mechanism to ensure we couldn't log in until the cloud-init is executed. We might not have planned for that to be the case, but that could be a dependency that has developed over time.

Discussed with @EdDev , we prefer this approach, decouple the modules and wait for each as needed (as you also suggested above).
https://github.com/kubevirt/kubevirtci/pull/550#issuecomment-785194507

File: tests/libvmi/factory.go
Comment (oshoval) [reply tests/libvmi/factory.go:55]: Basically it isn't needed, 
but from what i remember tests which use NetworkData do need this, else they won't run Cloud-init at all, 
so felt its better to keep it, i can verify it if needed.
Thought its safe to keep it and it will also allow easier modification of the UserData

**Comment by EdDev** [comment]
> so, with this new sriov image, is the password preconfigured?

It is the case already, the provisioning of the existing (and future) image included the password.
The only difference is, that when cloud-init starts again, it resets the default account (fedora) so there is a need to re-specify it again.
(and this is what the current change is suppose to fix)

> 
> If so, how are we handling tests that need to ensure that the cloud-init script has executed before peforming the console login... for example, if a cloud-init script is setting something up for a test, and the fedora login can happen before the cloud-init script executes, then the test might kick off before the initial setup has completed.

With the Fedora custom image, this was the case already IIUC.
A login could have occurred even before could-init started, as the account was set in the provision stage.

> 
> I think the fedora password change in the cloud init was actually acting as sort of a gating mechanism to ensure we couldn't log in until the cloud-init is executed. We might not have planned for that to be the case, but that could be a dependency that has developed over time.

This was answered in the context of solving the problem in a following change.
But I will take this opertunity to emphasize that IMO the root cause of all the pain is the need to access the VM though the console to execute operation. That should be the last resort and not the first solution.
All operations should be placed in cloud-init (even deploying services if needed) and the console should be preserved for the cases where we have no other choice (e.g. check results of operations).
One example is the IP configurations... we changed them from console based to cloud-init network based.

File: tests/libvmi/factory.go
Comment (EdDev) [inline tests/libvmi/factory.go:32]: Please separate from the above ones using a newline, these belong to different things.

**Comment by oshoval** [comment]
Addressed comments (see below what i didn't), and refactor according the offline suggestions.
Removed the empty UserData as its not needed after testing it, tests are passing locally.

Adding a check for Cloud-init done before setting ips in case needed will be done on a follow PR,
the password was not a barrier even before, as explained by Edy and myself, unless required otherwise.

**Comment by oshoval** [comment]
Rebase

File: WORKSPACE
Comment (RamLavi) [reply WORKSPACE:328]: which PR? this one?

Comment (oshoval) [reply WORKSPACE:328]: https://github.com/kubevirt/kubevirtci/pull/556 need to be lgtm approve + merge
and then the current PR https://github.com/kubevirt/kubevirt/pull/5135 need to be lgtm + approve + hold
and then i will change this snippet to an image which will be build due to https://github.com/kubevirt/kubevirtci/pull/556 
and pushed to quay.io/kubevirtci, and remove the hold

File: tests/libvmi/factory.go
Comment (RamLavi) [inline tests/libvmi/factory.go:55]: not sure I'm getting it.

the commit message suggests you're removing the userDataSetPassword
 from the configuration, but from what I understand, you've separated it to a different function `newFedoraWithUserDataSetPassword` and then send that to use that to `newFedora`. 
So eventually you **do** add it, but in a former function.. IIUC nothing changed really.. 
Am I missing something?

Comment (oshoval) [reply tests/libvmi/factory.go:55]: I removed it only from NewSriovFedora, the change you point to is to keep the non sriov intact

Comment (RamLavi) [reply tests/libvmi/factory.go:55]: yep you're right, must have missed it. then I'm good. thanks

Comment (RamLavi) [reply tests/libvmi/factory.go:55]: Can you also explain why this is an issue only with NewSriovFedora vms and this race does not also occur on regular ones?

Comment (oshoval) [reply tests/libvmi/factory.go:55]: the race do exists on the regular ones, the solution there atm is different 
https://github.com/kubevirt/kubevirt/pull/5019
It disable the Guest Agent service
install it as the last thing it does on Cloud-Init, after setting the password on Cloud-Init,
wait for the Guest Agent connected event,
this way it knows the password was configured

This PR suggest other solution and focus on the SRIOV first, (it has its own container disk atm anyhow)

Comment (RamLavi) [reply tests/libvmi/factory.go:55]: I understand. thanks!

**Review by RamLavi — APPROVED** [review]
/lgtm

**Comment by oshoval** [comment]
Thanks Ram
@EdDev please re-review when you can the two PRs so we can proceed
thanks

**Comment by oshoval** [comment]
/cc @AlonaKaplan

can you take a look please?

**Comment by davidvossel** [comment]
> In order to fix the race and keep the code as simple as possible,
>this PR does the following:
>
>Use a VM without Cloud-init module which reconfigure the fedora user
>to its default values.
>Remove the setting of the password via UserData, as it doesn't needed anymore,
>since the password now isn't reset by the Cloud-init default config.


@oshoval @EdDev 

How would you all feel about using the guest agent as the mechanism to indicate whether cloud-init (at least the user data portion) has completed.

With the FedoraTools pre-provisioned container disk, i pre-installed the installed the guest agent, but don't actually let the guest agent start at boot. The cloud-init user data starts the agent. It looks a little strange, but i have to move the systemd unit file around to prevent it from starting at boot, and then use the userData to re-install/start the agent.

```
func GetFedoraToolsGuestAgentUserData() string {
        return `#!/bin/bash
            echo "fedora" |passwd fedora --stdin
            sudo setenforce Permissive
            sudo cp /home/fedora/qemu-guest-agent.service /lib/systemd/system/
            sudo systemctl daemon-reload
            sudo systemctl start qemu-guest-agent
            sudo systemctl enable qemu-guest-agent
`
}

```


Now, after i do that, i can wait for the "AgentConnected" condition which indicates to me that everything in the userData has executed. If i structure my tests like this...

- create VMI
- wait for agent connected condition
- login to console

Then i never have to worry about the login failing because I'm certain the password has already been set. 

I trust whatever you all come up with here, just wanted to throw out an idea in case you hadn't considered it yet.

**Comment by oshoval** [comment]
> > In order to fix the race and keep the code as simple as possible,
> > this PR does the following:
> > Use a VM without Cloud-init module which reconfigure the fedora user
> > to its default values.
> > Remove the setting of the password via UserData, as it doesn't needed anymore,
> > since the password now isn't reset by the Cloud-init default config.
> 
> @oshoval @EdDev
> 
> How would you all feel about using the guest agent as the mechanism to indicate whether cloud-init (at least the user data portion) has completed.
> 
> With the FedoraTools pre-provisioned container disk, i pre-installed the installed the guest agent, but don't actually let the guest agent start at boot. The cloud-init user data starts the agent. It looks a little strange, but i have to move the systemd unit file around to prevent it from starting at boot, and then use the userData to re-install/start the agent.
> 
> ```
> func GetFedoraToolsGuestAgentUserData() string {
>         return `#!/bin/bash
>             echo "fedora" |passwd fedora --stdin
>             sudo setenforce Permissive
>             sudo cp /home/fedora/qemu-guest-agent.service /lib/systemd/system/
>             sudo systemctl daemon-reload
>             sudo systemctl start qemu-guest-agent
>             sudo systemctl enable qemu-guest-agent
> `
> }
> ```
> 
> Now, after i do that, i can wait for the "AgentConnected" condition which indicates to me that everything in the userData has executed. If i structure my tests like this...
> 
> * create VMI
> * wait for agent connected condition
> * login to console
> 
> Then i never have to worry about the login failing because I'm certain the password has already been set.
> 
> I trust whatever you all come up with here, just wanted to throw out an idea in case you hadn't considered it yet.

Thanks David, we considered that, 
see please https://github.com/kubevirt/kubevirtci/pull/550#issuecomment-785194507
We found this PR approach favorable, as it simplify things.

**Comment by EdDev** [comment]
> How would you all feel about using the guest agent as the mechanism to indicate whether cloud-init (at least the user data portion) has completed.
> 
> With the FedoraTools pre-provisioned container disk, i pre-installed the installed the guest agent, but don't actually let the guest agent start at boot. The cloud-init user data starts the agent. It looks a little strange, but i have to move the systemd unit file around to prevent it from starting at boot, and then use the userData to re-install/start the agent.
> 
> 
> Now, after i do that, i can wait for the "AgentConnected" condition which indicates to me that everything in the userData has executed. If i structure my tests like this...
> 
>     * create VMI
> 
>     * wait for agent connected condition
> 
>     * login to console
> 
> 
> Then i never have to worry about the login failing because I'm certain the password has already been set.
> 
> I trust whatever you all come up with here, just wanted to throw out an idea in case you hadn't considered it yet.

@davidvossel , thank you for pointing this out.
My main concern here is that we require cloud-init user data to be specified with this "hack" in place. This has implications for tests that want to add more actions and therefore need to override the existing user-data and make sure it includes the section you just mentioned. This was the case with the password, but I hope we can drop it and use some smother option.

What do you say about adjusting the guest-agent unit configuration to start after the cloud-init services? This is something that systemd should support.

There is an additional point which may be even stronger about this sequence of actions.
There is nothing that assures that the actions in the cloud-init user or network data are blocking/synchronous. That means that even if we want for the guest-agent signal, it says nothing about what we expected to see as a result of the actions in the cloud-init. This is especially relevant for networking, as dynamic IP allocations when set are not blocking and the service (NetworkManager) will wait in the background for the DHCP negotiation.

I'm not trying to say it is not useful to get the signal from the guest-agent last, but I am trying to say that it may not be enough.
This change only solves the race with the account configuration, it does not attempt to solve the problem where tests assume that some operation has been completed when the prompt is available.

File: tests/utils.go
Comment (EdDev) [inline tests/utils.go:2941]: Can you please explain how is this function useful when there is no login?
Seems to me that the function should just be removed and the `console.Login*` ones be used directly.

**Comment by oshoval** [comment]
> What do you say about adjusting the guest-agent unit configuration to start after the cloud-init services? This is something that systemd should support.

IMHO, its preferable to not create this type of dependency, keep it simple is favorable.
Moreover i tried it a bit and it didn't work out of the box, (of course i might have mistaken),
I don't think we need to mix the cloud-init dependency chain with other dependencies, we can see for example what happened 
with the password as a gating that started small and become bigger, and required maintenance.

Secondly it will enforce us to run cloud-init, where now we are not enforced to do so on the tests that doesn't use
NetworkData.
On any case, i would appreciate please if such a change, if needed, can be done on other PR, as this PR is self contained,
and fix a real problem without regression.

File: tests/libvmi/factory.go
Comment (EdDev) [inline tests/libvmi/factory.go:37]: This new `newFedoraWithUserDataSetPassword` is not needed, you should just have:
```
newFedora(cd.ContainerDiskFedora, WithCloudInitNoCloudUserData(userDataSetPassword, false), opts...)
```
Where `userDataSetPassword is a constant`.

**Review by EdDev — CHANGES_REQUESTED** [review]

**Comment by EdDev** [comment]
> > What do you say about adjusting the guest-agent unit configuration to start after the cloud-init services? This is something that systemd should support.
> 
> IMHO, its preferable to not create this type of dependency, keep it simple is favorable.
> Moreover i tried it a bit and it didn't work out of the box, (of course i might have mistaken),
> I don't think we need to mix the cloud-init dependency chain with other dependencies, we can see for example what happened
> with the password as a gating that started small and become bigger, and required maintenance.

I do not see a problem in running guest-agent after. I see no scenario in which it can hurt in any way, but if you do have a specific scenario, you are welcome to present it.
I did mentioned that is is not enough, but that is not enough to dismiss it.

> 
> Secondly it will enforce us to run cloud-init, where now we are not enforced to do so on the tests that doesn't use
> NetworkData.


I do not understand this point. could-init service must run, at least in order to check if there is a configuration file to execute (be it user or network data). And in any case, systemd has many options do deal with such things, I would be surprise to see this not working.

> On any case, i would appreciate please if such a change, if needed, can be done on other PR, as this PR is self contained,
> and fix a real problem without regression.

File: tests/utils.go
Comment (oshoval) [reply tests/utils.go:2941]: It just nice to have, not related / nor mandatory for this PR.
I will drop this commit from this PR, its not in the scope.
its was just in order to be able to call `WaitUntilVMIReadyWithContext` family without giving a LoginToFactory

File: tests/libvmi/factory.go
Comment (oshoval) [reply tests/libvmi/factory.go:37]:
```
too many arguments in call to newFedora
	have (tests.ContainerDisk, Option, []Option...)
	want (tests.ContainerDisk, ...Option)go
```
	
I can append it in place but then i will need to do it twice (this is why i added a function),
or if there is a more elegant way ? but specifically what you suggested gives an error.

Comment (EdDev) [reply tests/libvmi/factory.go:37]: Ahh, right. So just append it in a one-liner before.
What exactly you need to do twice?

Comment (oshoval) [reply tests/libvmi/factory.go:37]: do this append twice, because there are two functions that need it
NewFedora
NewTestToolingFedora

**Comment by oshoval** [comment]
Addressed comments,
did the UserData cosmetics in a separate commit, so i can revert easily if needed, and squash it looks good

**Comment by EdDev** [comment]
> Addressed comments,
> did the UserData cosmetics in a separate commit, so i can revert easily if needed, and squash it looks good

LGTM, please squash.
Worth mentioning in the commit message why you add the login option at the head, it may not be that trivial to a reader.

**Comment by oshoval** [comment]
Addressed comments.

**Review by EdDev — APPROVED** [review]
Thanks, looks good.
We will need to wait for https://github.com/kubevirt/kubevirtci/pull/556 to get in, push to the registry the new image,  change the first commit to point to the new image and finally let this in.

Hope I got this right.

Can you just confirm that this will fail CI in case the Fedora image used is the old one?

**Comment by oshoval** [comment]
> Thanks, looks good.
> We will need to wait for [kubevirt/kubevirtci#556](https://github.com/kubevirt/kubevirtci/pull/556) to get in, push to the registry the new image, change the first commit to point to the new image and finally let this in.
> 
> Hope I got this right.

You got it right, beside the nit below.

> 
> Can you just confirm that this will fail CI in case the Fedora image used is the old one?

It will not fail CI if the image is the old one, because if there is no UserData (as this PR remove the UserData from SRIOV VM), cloud init doesn't run (check it),
and since its not running it won't run the module which removes the password.

**Comment by EdDev** [comment]
> It will not fail CI if the image is the old one, because if there is no UserData (as this PR remove the UserData from SRIOV VM), cloud init doesn't run (verified it),
> and since its not running it won't run the module which removes the password.

The you should make it fail, otherwise we do not know if the change took affect or not.
One way could be to remove the password configuration from one of the tests that use it. Can you locate such a test in the network suite?

**Comment by oshoval** [comment]
> > It will not fail CI if the image is the old one, because if there is no UserData (as this PR remove the UserData from SRIOV VM), cloud init doesn't run (verified it),
> > and since its not running it won't run the module which removes the password.
> 
> Then you should make it fail, otherwise we do not know if the change took effect or not.
> One way could be to remove the password configuration from one of the tests that use it. Can you locate such a test in the network suite?

We can know its the image that runs, by just checking after its merged that the cloud.cfg doesn't have this line, i verified it back and forth.
Before we over engineer lets talk exactly on the scenario that would happen, because the previous info might have some side branches.

1. https://github.com/kubevirt/kubevirtci/pull/556 is merged, if its merged with the same tag as the current, new PRs would totally fail because they didnt update the sha256 and won't manage to download the image.
If it merged with other name, the images will leave side by side.
which of this situations interest you ?
from there lets continue please

I prefer lets take it offline if possible, because its a long discussion.

**Comment by EdDev** [comment]
> > > It will not fail CI if the image is the old one, because if there is no UserData (as this PR remove the UserData from SRIOV VM), cloud init doesn't run (verified it),
> > > and since its not running it won't run the module which removes the password.
> > 
> > 
> > Then you should make it fail, otherwise we do not know if the change took effect or not.
> > One way could be to remove the password configuration from one of the tests that use it. Can you locate such a test in the network suite?
> 
> We can know its the image that runs, by just checking after its merged that the cloud.cfg doesn't have this line, i verified it back and forth.
> Before we over engineer lets talk exactly on the scenario that would happen, because the previous info might have some side branches.
> 
>     1. [kubevirt/kubevirtci#556](https://github.com/kubevirt/kubevirtci/pull/556) is merged, if its merged with the same tag as the current, new PRs would totally fail because they didnt update the sha256 and won't manage to download the image.
>        If it merged with other name, the images will leave side by side.
>        which of this situations interest you ?
>        from there lets continue please

What tag are yo talking about? kubevirt/kubevirt is pointing to a SHA, not to a tag.

I asked for this PR to fail if it is using the current Fedora SRIOV VM image and not the proposed one with the changes.

> 
> 
> EDIT:
> if we push the same image_name:tag (which what i believe we will do), PRs that don't point to the new hash, will have a mismatch, causing the whole job to fail, and it might already do what you asked, wdyt?

You explained [here](https://github.com/kubevirt/kubevirt/pull/5135#issuecomment-799310937) that it will not fail.
If it fails, you answered what I asked and I do not understand what you argued against.

**Comment by oshoval** [comment]
> > > > It will not fail CI if the image is the old one, because if there is no UserData (as this PR remove the UserData from SRIOV VM), cloud init doesn't run (verified it),
> > > > and since its not running it won't run the module which removes the password.
> > > 
> > > 
> > > Then you should make it fail, otherwise we do not know if the change took effect or not.
> > > One way could be to remove the password configuration from one of the tests that use it. Can you locate such a test in the network suite?
> > 
> > 
> > We can know its the image that runs, by just checking after its merged that the cloud.cfg doesn't have this line, i verified it back and forth.
> > Before we over engineer lets talk exactly on the scenario that would happen, because the previous info might have some side branches.
> > ```
> > 1. [kubevirt/kubevirtci#556](https://github.com/kubevirt/kubevirtci/pull/556) is merged, if its merged with the same tag as the current, new PRs would totally fail because they didnt update the sha256 and won't manage to download the image.
> >    If it merged with other name, the images will leave side by side.
> >    which of this situations interest you ?
> >    from there lets continue please
> > ```
> 
> What tag are yo talking about? kubevirt/kubevirt is pointing to a SHA, not to a tag.

This [1] is the hash that point to the VM container image,
if:
1. we push the new image instead of the current Sriov VM container image
(since we change the image, its sha256 will be changed as well, the same one that i talk about in [2] below)
2. we don't change the VM sha256 in kubevirt, (as you asked, if we still point to the old one)
https://github.com/kubevirt/kubevirt/pull/5135/files#diff-5493ff8e9397811510e780de47c57abb70137f1afe85d1519130dc3679d60ce5L326
The whole lane will fail `cluster-sync` because it won't find an image that correspond to the old hash.

> 
> I asked for this PR to fail if it is using the current Fedora SRIOV VM image and not the proposed one with the changes.
> 
> > EDIT:
> > if we push the same image_name:tag (which what i believe we will do), PRs that don't point to the new hash, will have a mismatch, causing the whole job to fail, and it might already do what you asked, wdyt?
> 
> You explained [here](https://github.com/kubevirt/kubevirt/pull/5135#issuecomment-799310937) that it will not fail.
> If it fails, you answered what I asked and I do not understand what you argued against.

Disregard what i explained before, its a specific case that we won't reach because the above scenario of this comment.

**Comment by EdDev** [comment]
> > What tag are yo talking about? kubevirt/kubevirt is pointing to a SHA, not to a tag.
> 
> This [A] is the hash that point to the VM container image.
> 
> If the conditions below are met, the lane would fail:
> 
>     1. we push the new image instead of the current Sriov VM container image
>        (since we change the image, its sha256 will be changed as well, the same one that i talk about in [2] below)
> 
>     2. we don't change the VM sha256 in kubevirt, (as you asked, if we still point to the old one)
>        [A] https://github.com/kubevirt/kubevirt/pull/5135/files#diff-5493ff8e9397811510e780de47c57abb70137f1afe85d1519130dc3679d60ce5L326
>        The whole lane will fail `cluster-sync` because it won't find an image that correspond to the old hash.

That is not how it works. The digest SHA is unique, it cannot be replaced by another and it will stay in the registry until someone prunes it. This is how kubevirtci and its consumers work at the moment and depend on this behavior.
Tags point to a SHA, so it is possible to use the same tag and point to another SHA, but this is not the case here.

> 
> 
> EDIT: made sure.
> As long as we replace the image (i.e push it as "latest"), and not publish it side by side with the previous one,
> `make cluster-sync` would failed because it can't find the hash here
> https://quay.io/repository/kubevirtci/fedora-sriov-testing?tag=latest&tab=tags

Sorry, I do not follow this comment. We do not use any tag (incluging "latest"), we use a SHA which is assumed to be always there, even if you push 10 more images.


If this PR will not fail without the first commit now, we will not be able to trust it.
This change is adding a new behavior and updating the VMI creation helpers (by removing the default cloud-init user data with the login). If everything still works without the behavior change in the image, you cannot reason for the change in it at all.

**Comment by oshoval** [comment]
> That is not how it works. The digest SHA is unique, it cannot be replaced by another and it will stay in the registry until someone prunes it. This is how kubevirtci and its consumers work at the moment and depend on this behavior.
> Tags point to a SHA, so it is possible to use the same tag and point to another SHA, but this is not the case here.

Please create your own quay.io and push as latest,
you will see its prune the previous sha256, because it wasn't tagged with another tag,
https://quay.io/repository/kubevirtci/fedora-sriov-testing?tag=latest&tab=tags
See here there is only one sha256
Hence it will fail because it will search the previous sha256 and won't find it, it was pruned
We had this problem in the past with ocp providers, the previous latest was pruned,
so we fixed it by tagging them with an additional tag.

I just validated what i wrote
take an image, push it, save the hash of it in your notebood
[1] docker pull quay.io/oshoval/try@sha256:32150bb4704c0f9cf10599cb7fd7165841f73f1dfe4a96243e5e20e54f66e319

then push other image with the same tag (latest)
and see it prune the previous,
you can see also on quay.io its prune it
```
~ $ docker push quay.io/oshoval/try
The push refers to repository [quay.io/oshoval/try]
18e5c7523dfc: Pushed 
1cce3be9b3dc: Pushed 
ab7d89ca58af: Pushed 
a008320355ce: Pushed 
6169bf830ae6: Pushed 
latest: digest: sha256:32150bb4704c0f9cf10599cb7fd7165841f73f1dfe4a96243e5e20e54f66e319 size: 1363
~ $ docker pull quay.io/oshoval/try@sha256:32150bb4704c0f9cf10599cb7fd7165841f73f1dfe4a96243e5e20e54f66e319
sha256:32150bb4704c0f9cf10599cb7fd7165841f73f1dfe4a96243e5e20e54f66e319: Pulling from oshoval/try
Digest: sha256:32150bb4704c0f9cf10599cb7fd7165841f73f1dfe4a96243e5e20e54f66e319
Status: Image is up to date for quay.io/oshoval/try@sha256:32150bb4704c0f9cf10599cb7fd7165841f73f1dfe4a96243e5e20e54f66e319
~ $ 
```
```
~ $ docker tag quay.io/oshoval/fedora:32 quay.io/oshoval/try
~ $ docker push quay.io/oshoval/try
The push refers to repository [quay.io/oshoval/try]
162dea22b0e9: Mounted from oshoval/fedora 
latest: digest: sha256:ce793f24faa67ea021ace78a736cd331208e28bc26ec212d429cf93560b33208 size: 529
~ $ docker pull quay.io/oshoval/try@sha256:32150bb4704c0f9cf10599cb7fd7165841f73f1dfe4a96243e5e20e54f66e319
Error response from daemon: manifest for quay.io/oshoval/try@sha256:32150bb4704c0f9cf10599cb7fd7165841f73f1dfe4a96243e5e20e54f66e319 not found: manifest unknown: manifest unknown
```

**Comment by EdDev** [comment]
Summarizing this long discussion:

On Quay registry, unlike the previous DockerHub registry, individual digests are not preserved if there is no TAG associated with them.
Therefore, in this case it will indeed fail just because we use no tags when pushing (needs to be fixed).

With all that clarified, I expect this change to fail if running with the current image (i.e. removing the first commit). If it does not, we have a problem.

**Comment by oshoval** [comment]
Updated first commit to point to the new image
the new image was pushed side by side, 
verified locally that the new image is fetched (by login to the VM, and checking the provision date + the cloud.cfg which is the file that was changed when the new image is built)

/hold cancel
the PR this PR depended on, was merged

**Comment by oshoval** [comment]
Rebase

**Comment by EdDev** [comment]
> verified locally that the new image is fetched (by login to the VM, and checking the provision date + the cloud.cfg which is the file that was changed when the new image is built)

Is this PR failing relevant tests without updating the VM image?

**Comment by oshoval** [comment]
> > verified locally that the new image is fetched (by login to the VM, and checking the provision date + the cloud.cfg which is the file that was changed when the new image is built)
> 
> Is this PR failing relevant tests without updating the VM image?

I think it won't fail , because this PR removes the UserData which is a partial workaround of the problem that this PR intend to fix.
I don't think we should care about it much, the image that is fetched is for sure the one according the hash.
We can create other PR that will tag the images with some string when it builds them (for example the date of the build + the commit of the build), that file will exists in the VM, and the test will check if the file exists there and have the expected content,
alternativly we can just check the date of the /var/lib/cloud/instance/boot-finished when no UserData nor NetworkData exists, and see its 17.3 (which is what i did manually), and with the last image it was other date.
But imo we are over engineering, and double checking because the hash is our guard, it won't take any other image.
Please select what action item u suggest.

**Comment by EdDev** [comment]
> > > verified locally that the new image is fetched (by login to the VM, and checking the provision date + the cloud.cfg which is the file that was changed when the new image is built)
> > 
> > 
> > Is this PR failing relevant tests without updating the VM image?
> 
> Reminder - we didn't push the image instead the previous one, see below please the consequences.
> 
> I think it won't fail , because this PR removes the UserData which is a partial workaround of the problem that this PR intend to fix.
> I don't think we should care about it much, the image that is fetched is for sure the one according the hash.
> We can create other PR that will tag the images with some string when it builds them (for example the date of the build + the commit of the build), that file will exists in the VM, and the test will check if the file exists there and have the expected content,
> alternativly we can just check the date of the /var/lib/cloud/instance/boot-finished when no UserData nor NetworkData exists, and see its 17.3 (which is what i did manually), and with the last image it was other date.
> But imo we are over engineering, and double checking because the hash is our guard, it won't take any other image.
> Please select what action item u suggest.

I am not interested in confirming we took the right image or not, that should be transparent if the functionality fails without the enhanced new image.

Why would we take the new image if everything works without it? This is the core issue in hand.

**Comment by oshoval** [comment]
> I am not interested in confirming we took the right image or not, that should be transparent if the functionality fails without the enhanced new image.
> 
> Why would we take the new image if everything works without it? This is the core issue in hand.

It works but its flaky, this what i mean.
because there is a race, it works until it doesn't.
This PR fix this race.

**Comment by EdDev** [comment]
> > I am not interested in confirming we took the right image or not, that should be transparent if the functionality fails without the enhanced new image.
> > Why would we take the new image if everything works without it? This is the core issue in hand.
> 
> It works but its flaky, this what i mean.
> because there is a race, it works until it doesn't.
> This PR fix this race.

Thank you for clarifying this point. It was missed in between all the other discussions.
I think the source of the confusion here is that this PR tries to do two unrelated things that could have been totally split. Then it would have been clear one is not related to the other (removing the default could-init, adjusting the image).

**Review by EdDev — APPROVED** [review]
Thank you.

**Comment by oshoval** [comment]
Thanks

/retest pull-kubevirt-check-tests-for-flakes

seem the job is broken a bit

**Comment by EdDev** [comment]
> Thanks
> 
> /retest pull-kubevirt-check-tests-for-flakes
> 
> seem the job is broken a bit

It is not a required job, why do you need to bother?

**Comment by oshoval** [comment]
> > Thanks
> > /retest pull-kubevirt-check-tests-for-flakes
> > seem the job is broken a bit
> 
> It is not a required job, why do you need to bother?

You right, i am used that it was required i think, anyhow notified about it.

**Pull request merged**
